gem/refactorise.py

```diff
@@ -112,6 +116,135 @@ def product(*args):
+
+        # constraints
+        for connection in connections:
+            ilp_prob += ilp.lpSum(ilp_var[index] for index in connection) >= 1
+
+        ilp_prob.solve()
+        if ilp_prob.status != 1:
+            raise AssertionError("Something bad happened during ILP")
+
+        optimal_atomics = [atomic for atomic, _index in iteritems(atomic_index) if ilp_var[_index].value() == 1]
+        other_atomics = [atomic for atomic, _index in iteritems(atomic_index) if ilp_var[_index].value() == 0]
+        optimal_atomics = sorted(optimal_atomics, key=lambda x: self.argument_indices_extent(x), reverse=True)
+        other_atomics = sorted(other_atomics, key=lambda x: self.argument_indices_extent(x), reverse=True)
+        return (tuple(optimal_atomics), tuple(other_atomics))
+
+    def factorise_atomics(self, optimal_atomics):
+        if not optimal_atomics:
+            return
+        if len(self.ordering) < 2:
+            return
+        # pick the first atomic
+        (sum_indices_set, sum_indices), atomic = optimal_atomics[0]
+        factorised_out = []
+        for key, (_, _atomics) in iteritems(self.ordering):
+            if sum_indices_set == key[0] and atomic in _atomics:
+                factorised_out.append(key)
+        if len(factorised_out) <= 1:
+            self.factorise_atomics(optimal_atomics[1:])
+            return
+        all_atomics = []
+        all_rest = []
+        for key in factorised_out:
+            _atomics = list(self.ordering[key][1])
+            _atomics.remove(atomic)
+            all_atomics.append(_atomics)
+            all_rest.append(self.monomials[key])
+            del self.ordering[key]
+            del self.monomials[key]
+        new_monomial_sum = MonomialSum()
+        new_monomial_sum.argument_indices = self.argument_indices
+        for _atomics, _rest in zip(all_atomics, all_rest):
+            new_monomial_sum.add((), _atomics, _rest)
+        new_optimal_atomics = [((frozenset(), ()), oa) for _, oa in optimal_atomics[1:]]
+        new_monomial_sum.factorise_atomics(new_optimal_atomics)
+        assert len(new_monomial_sum.ordering) != 0
+        if len(new_monomial_sum.ordering) == 1:
+            # result is a product
+            new_atomics = list(itervalues(new_monomial_sum.ordering))[0][1] + (atomic,)
+            new_rest = list(itervalues(new_monomial_sum.monomials))[0]
+        else:
+            # result is a sum
+            new_node = new_monomial_sum.to_expression()
+            new_atomics = [atomic]
+            new_rest = one
+            if set(self.argument_indices) & set(new_node.free_indices):
+                new_atomics.append(new_node)
+            else:
+                new_rest = new_node
+        self.add(sum_indices, new_atomics, new_rest)
+        # factorise the next atomic
+        self.factorise_atomics(optimal_atomics[1:])
+        return
```

Comment:
The new version is (marginally) better when I wrap the logic inside MonomialSum.optimise(). I think I can rewrite this later, as each atomic factor is applied to different monomials in the sum, so that part doesn't need to be recursive (only the factorised parts need to be recursive, but it is on a new MonomialSum object)